### PR TITLE
Tweak parallelism from A/B config

### LINKED ISSUE #483: Configure number of workers and instance types in A/B tests
At the moment, all of our tests use hardcoded number of VMs and instance types - 10 x m6i.large for all tests that use the `small_client` fixture.

This leaves us blind to three major use cases:
- workers with more than 2 threads
- clusters with 20+ workers - this triggers completely different code paths in the scheduler
- clusters with many workers (100~10,000) - this can highlight problems with big-O complexity in the scheduler algorithms.

# Proposed design
It would be extremely costly to enable the above in the nightly tests, so I propose to add a switch to A/B tests and trigger them as needed.

Require new config files `AB_environments/AB_<name>.cluster.yaml`, with the following config:
```yaml
# These are arbitrary kwargs to coiled.Cluster

# For all tests using the small_client fixture
small_cluster:
  n_workers: 10
  worker_vm_types: [m6i.large]  # 2CPU, 8GiB
  scheduler_vm_types: [m6i.xlarge]

# For test_parquet.py
parquet_cluster:
  n_workers: 15
  worker_vm_types: [m5.xlarge]  # 4 CPU, 16 GiB
  scheduler_vm_types: [m5.xlarge]
```

and have them reach the respective fixtures.
Note that the above will not impact special tests which create their own clusters, e.g. test_spill.py.

Review all tests to verify that the test workload scales with the total amount of CPUs or memory in the cluster, so that you get a meaningful workload on a 10 workers cluster as well as on a 200 workers one.

Review which tests hit the spill threshold and, 
  - if spilling isn't an integral part of the test, make sure the test data is sized so that spilling is consistently avoided across various cluster sizes and threads per worker
  - (harder) if spilling IS an integral part of the test (read: shuffle), make sure the test data is sized so that spilling is consistently hit in roughly the same proportion

Note that, even with the review in dataset scaling above, it is expected that testing e.g. a 20-threads cluster vs. a 40-threads cluster is going to be comparing apples and oranges - namely for all tests that implement an algorithm that doesn't scale linearly with the number of chunks. However it will be very interesting to compare eg. how  a 10 workers, 2 threads-per-worker cluster compares to a 5 workers, 4 threads-per-worker one.

### PR DESCRIPTION
**Pull request by crusaderky**
- Partially closes #483 

When one significantly increases the cluster size (#549), the current hardcoded levels of parallelism can cause Coiled to fail to allocate all the necessary workers.

Additionally, pytest-xdist tends to eat the pytest output, making debugging more difficult.
Expose parallelisation levels as knobs in the A/B config file and optionally disable pytest-xdist completely.

### CODE REVIEW

**Comment by ntabris** [comment]
What sort of cluster size are you talking about? Do you have example of test runs where Coiled failed to provision the desired number of instances?

**Comment by crusaderky** [comment]
@ntabris in https://github.com/coiled/coiled-runtime/pull/355 we ascertained that AWS DoS protection will kick in beyond a certain, unspecified threshold.

#483 is about having the option to play with 1,000 workers clusters.
This PR makes the difference between 1,000 instances at once and 20,000.

**Comment by ntabris** [comment]
I don't know what you mean by AWS DoS, there are lots of limits and I don't see anything in #355 that describes what you were seeing.

Maybe let's talk before you start making multiple 1000 worker clusters. That's fine to do occasionally, but we should make sure we aren't doing this way more than necessary, and 20 1000 worker clusters is a lot (even if you're making them one at a time).

**Comment by crusaderky** [comment]
My above comment also applies to 100 workers. This PR makes the difference between 100 workers and 2000 workers.

**Comment by ntabris** [comment]
Sure, I'm not saying this PR has a problem, I'm just saying me/platform should be in the loop when you're thinking about doing larger scale testing.

File: AB_environments/config.yaml [hunk 1]
```diff
@@ -14,3 +14,16 @@ categories:
...
+# create the Coiled clusters, and restarting failed jobs won't fix the problem.
+# Additionally, there are problems with Coiled itself triggered by limitations that are
+# never actually reached with real paying users.
+max_parallel:
+  # Number of parallel A/B test jobs per branch.
+  ci_jobs: 5
+  # Number of parallel test_*.py modules per A/B test job.
+  # Each module typically spawns one Coiled cluster at a time.
+  # Set to 1 to disable pytest-xdist.
+  modules_per_job: 4
```
Comment (hendrikmakait) [inline AB_environments/config.yaml:29]: nit: I first ran into the variable name without the description and it took me a moment to figure out how the number of modules would impact parallelism. Would it make sense to rename this to something like `pytest_workers_per_job` to highlight that this is about `pytest-xdist` parallelism?

File: AB_environments/README.md [hunk 2]
```diff
@@ -154,6 +154,9 @@ categories:
   - runtime
   - benchmarks
   - stability
+max_parallel:
```
Comment (ncclementi) [inline AB_environments/README.md:157]: The same way we explain opther parameters (e.g category, test_null_hypothesis, etc ) in the README, can we explain these two in the README as well? 

I know there are some comments in the `config.yml` but it'll be useful having them explain in the readme too.

**Review by ncclementi — APPROVED** [review]
LGTM, just left a small comment

File: AB_environments/config.yaml (hunk 1 above)
Comment (crusaderky) [reply AB_environments/config.yaml]: done

File: AB_environments/README.md (hunk 2 above)
Comment (crusaderky) [reply AB_environments/README.md:163]: done

**Pull request merged**